### PR TITLE
Tdrd 163 metadata validation repo schema creation

### PR DESCRIPTION
**Pull request by ian-hoyle**
This is the creation of the repo and initial ideas for the schema. The Schema will be expanded as requirements are confirmed. 
A meta schema will be useful to allow automatic type conversion and field key mapping
Example validation usage can be found in the Scala tests.

### CODE REVIEW

**Review by TomJKing — COMMENTED** [review]
Looks good. Thanks. Mostly minor comment and questions.

The unit test need to be made more comprehensive. It would be best to raise a new Jira task for that, as they should really cover all the schema rules which is a large task to do.

Will also need to add Github actions

**Comment by TomJKing** [comment]
@ian-hoyle / @TNAemmaPea This is a very large PR with a lot of changes. I would prefer if we can get it in as it is without addressing some of the comments from Emma at this stage. The schemas can be adjusted in later PRs as needed once full requirements are made clearer.

@ian-hoyle Can you address my PR comments and not make anymore changes once those are dealt with, so we can at least getting a draft of the schemas in. We can then also look at adding the necessary Github actions in another PR

File: metadata-schema/baseSchema.schema.json [hunk 1]
```diff
@@ -0,0 +1,350 @@
...
+        }
+      ],
+      "tdrDescription": null
+    },
+    "UUID": {
+      "type": "string",
+      "format": "uuid",
+      "alternateKeys": [
+        {
+          "tdrFileHeader": "FileReference"
```
Comment (TomJKing) [inline metadata-schema/baseSchema.schema.json:21]: The `tdrFileHeader` should be `UUID` not `FileReference`

File: metadata-schema/baseSchema.schema.json [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+        }
+      ],
+      "message": {
+        "format": "uuid must be a valid UUID"
+      },
+      "Emma Says": "This is a system generated field, not sure what we should say back to users if this fails"
```
Comment (TomJKing) [inline metadata-schema/baseSchema.schema.json:27]: Can you remove the `Emma Says` from the schema and either add them a comments or record them somewhere more appropriate.

**Comment by ian-hoyle** [comment]
_The unit test need to be made more comprehensive. It would be best to raise a new Jira task for that, as they should really cover all the schema rules which is a large task to do._  
The Scala code was added as a place to test JsonSchema validation tool to prove JSON schema could be programmatically used for validation of TNA metadata . This repo should really just hold the schema and all tests and usage places in the relevant metadata validator projects (possibly linked to from this repo).
A ticket will be added to remove the code and update the repo to provide greater description of how to add new fields and the use of schema language extensions such as alternateKeys

File: README.md [hunk 3]
```diff
@@ -1 +1,192 @@
...
+     "type": [
+        "string",
+        "null"
+     ],
+     "pattern": "^(0?[1-9]|[12][0-9]|3[01])[\\/\\-](0?[1-9]|1[012])[\\/\\-]\\d{4}$|\\d{4}-[01]\\d-[0-3]\\dT[0-2]\\d:[0-5]\\d:[0-5]\\d(?:\\.\\d+)?Z?",
+     "tdrName": "ClientSideFileLastModifiedDate"
+ }
+.....
+```
+Definition for a UUID
```
[the remark this answers is not in the document]
Comment (TomJKing) [reply README.md:63]: Please wrap the UUID as code.

File: README.md [hunk 4]
```diff
@@ -1 +1,192 @@
...
+* UUID - field key 
+* type - the value must be a string
+* format - the string format must be a uuid
+* tdrDescription - the human-readable key 
+* tdrName - alternate name for the key
+* message - the message used in error reporting when validating data
+
+Definition for date_last_modified
```
[the remark this answers is not in the document]
Comment (TomJKing) [reply README.md:70]: Please wrap the "date_last_modified" as code

File: README.md [hunk 5]
```diff
@@ -1 +1,192 @@
...
+* date_last_modified field key
+* type - value can be a string or null
+* pattern - the string will be in the format dd/mm/yyyy or yyyy-mm-ddThh:mm:ss
+* tdrName - alternate name used for key ClientSideFileLastModifiedDate
+
+Example data
+```angular2html
+{
+  "UUID": "f373d856-d4ee-4b41-ae89-d7327915c73e",
+  "file_path":"file:///E:/DADRI_001/content/windsor.pdf",
```
[the remark this answers is not in the document]
Comment (TomJKing) [reply README.md:80]: Is this file path real? If so it needs to be changed to something made up

File: README.md [hunk 6]
```diff
@@ -1 +1,192 @@
...
+* if closure_type OPEN
+  * Then
+    * closure_period must not be set or 0
+    * no closure_start_date
+    * ....
+  * else
+    * there must be a foi_exemption_asserted date
+    * the closure_type must be one of closed_review, closed_for or CLOSED
+    * ...
+    * there must be values for closure_period, closure_start_date, description_closed, foi_exemption_asserted, foi_exemption_code
```
[the remark this answers is not in the document]
Comment (TomJKing) [reply README.md:152]: Add a line break

File: README.md [hunk 7]
```diff
@@ -1 +1,192 @@
...
+Example data
+```angular2html
+{
+  "UUID": "f373d856-d4ee-4b41-ae89-d7327915c73e",
+  "file_path":"file:///E:/DADRI_001/content/windsor.pdf",
+  "foi_exemption_code":[],
+  "date_last_modified": "12/12/2001",
+  "description": "description for catalogue"
+}
+```
```
[the remark this answers is not in the document]
Comment (TomJKing) [reply README.md:85]: Add line break before the heading.

File: README.md [hunk 8]
```diff
@@ -1 +1,192 @@
-# da-metadata-schema
\ No newline at end of file
+# DA Metadata JSON Schema
```
[the remark this answers is not in the document]
Comment (TomJKing) [reply README.md:1]: Put DA out in full

File: build.sbt [hunk 9]
```diff
@@ -1,8 +1,68 @@
-ThisBuild / version := "0.1.0-SNAPSHOT"
+import Dependencies.*
+import sbt.url
+import sbtrelease.ReleaseStateTransformations._
 
-ThisBuild / scalaVersion := "3.3.3"
+ThisBuild / organization := "uk.gov.nationalarchives"
+ThisBuild / organizationName := "National Archives"
```
[the remark this answers is not in the document]
Comment (TomJKing) [reply build.sbt:6]: Put the organisation as "The National Archives"

File: build.sbt [hunk 10]
```diff
@@ -1,8 +1,68 @@
...
+  setReleaseVersion,
+  commitReleaseVersion,
+  tagRelease,
+  releaseStepCommand("publishSigned"),
+  releaseStepCommand("sonatypeBundleRelease"),
+  setNextVersion,
+  commitNextVersion,
+  pushChanges
+)
+
```
[the remark this answers is not in the document]
Comment (TomJKing) [reply build.sbt:49]: Remove unnecessary line breaks

File: metadata-schema/baseSchema.schema.json [hunk 11]
```diff
@@ -0,0 +1,307 @@
...
+        "string",
+        "null"
+      ],
+      "tdrName": "SHA256ClientSideChecksum",
+      "tdrDescription": null
+    }
+  }
+}
+
+
```
[the remark this answers is not in the document]
Comment (TomJKing) [reply metadata-schema/baseSchema.schema.json:357]: Remove unnecessary line breaks

File: metadata-schema/closureSchema.schema.json [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+          "foi_exemption_asserted": {
+            "type": "string",
+            "pattern": "^(0?[1-9]|[12][0-9]|3[01])[\\/\\-](0?[1-9]|1[012])[\\/\\-]\\d{4}$|\\d{4}-[01]\\d-[0-3]\\dT[0-2]\\d:[0-5]\\d:[0-5]\\d(?:\\.\\d+)?Z?"
+          },
+          "foi_exemption_code": {
+            "type": "array",
+            "minItems": 1,
+            "items": {
+              "type": "string",
+              "enum": [
```
[the remark this answers is not in the document]
Comment (TomJKing) [reply metadata-schema/closureSchema.schema.json:58]: Is this possible to define the the enum separately across the schemas?

File: metadata-schema/closureSchema.schema.json [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+                "EIR 12(5)(d)",
+                "EIR 12(5)(e)",
+                "EIR 12(5)(f)",
+                "EIR 12(5)(g)"
+              ]
+            }
+          },
+          "closure_type": {
+            "enum": [
+              "closed_review",
```
[the remark this answers is not in the document]
Comment (TomJKing) [reply metadata-schema/closureSchema.schema.json:104]: Is the above comment correct?

File: metadata-schema/closureSchema.schema.json [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+          "closure_start_date",
+          "description_closed",
+          "foi_exemption_asserted",
+          "foi_exemption_code",
+          "title_closed"
+        ]
+      }
+    }
+  ]
+}
```
[the remark this answers is not in the document]
Comment (TomJKing) [reply metadata-schema/closureSchema.schema.json:144]: Add missing line break

File: metadata-schema/relationshipSchema.schema.json [hunk 15]
```diff
@@ -0,0 +1,25 @@
+{
```
[the remark this answers is not in the document]
Comment (TomJKing) [reply metadata-schema/relationshipSchema.schema.json:1]: Is the above comment correct?

File: src/test/scala/uk/gov/tna/tdr/metadata/schema/validator/SchemaDataTypeSpec.scala [hunk 16]
```diff
@@ -0,0 +1,144 @@
+package uk.gov.tna.tdr.metadata.schema.validator
```
[the remark this answers is not in the document]
Comment (TomJKing) [reply src/test/scala/uk/gov/tna/tdr/metadata/schema/validator/SchemaDataTypeSpec.scala:1]: Remove tdr from the package  name

File: src/test/scala/uk/gov/tna/tdr/metadata/schema/validator/SchemaDataTypeSpec.scala [hunk 17]
```diff
@@ -0,0 +1,144 @@
...
+
+import com.fasterxml.jackson.databind.{JsonNode, ObjectMapper}
+import com.networknt.schema._
+import org.scalatest.matchers.should.Matchers._
+import org.scalatest.wordspec.AnyWordSpec
+
+import java.io.{File, InputStream}
+import java.nio.file.{Files, Path}
+import java.util
+
```
[the remark this answers is not in the document]
Comment (TomJKing) [reply src/test/scala/uk/gov/tna/tdr/metadata/schema/validator/SchemaDataTypeSpec.scala:11]: remove unncessary line break

File: src/test/scala/uk/gov/tna/tdr/metadata/schema/validator/SchemaDataTypeSpec.scala [hunk 18]
```diff
@@ -0,0 +1,144 @@
...
+
+class SchemaDataTypeSpec extends AnyWordSpec {
+
+  "schema validation" should {
+    "fail if no translation language when using relationshipSchema" in {
```
[the remark this answers is not in the document]
Comment (TomJKing) [reply src/test/scala/uk/gov/tna/tdr/metadata/schema/validator/SchemaDataTypeSpec.scala:16]: Rename the test

File: src/test/scala/uk/gov/tna/tdr/metadata/schema/validator/SchemaDataTypeSpec.scala [hunk 19]
```diff
@@ -0,0 +1,144 @@
...
+      val schema = getJsonSchemaFromStreamContentV7(schemaInputStream)
+      val dataInputStream = getClass.getResourceAsStream(dataPath)
+      val node = getJsonNodeFromStreamContent(dataInputStream)
+
+      val errors: util.Set[ValidationMessage] = schema.validate(node.toPrettyString, InputFormat.JSON)
+
+      errors.iterator().next().getMessage shouldBe "$: required property 'file_name_translation_language' not found"
+    }
+
+    "fail if no exemption code and document closed when using closureSchema " in {
```
[the remark this answers is not in the document]
Comment (TomJKing) [reply src/test/scala/uk/gov/tna/tdr/metadata/schema/validator/SchemaDataTypeSpec.scala:29]: Rename the test

File: src/test/scala/uk/gov/tna/tdr/metadata/schema/validator/SchemaDataTypeSpec.scala [hunk 20]
```diff
@@ -0,0 +1,144 @@
...
+      val schema = getJsonSchemaFromStreamContentV7(schemaInputStream)
+      val dataInputStream = getClass.getResourceAsStream(dataPath)
+      val node = getJsonNodeFromStreamContent(dataInputStream)
+
+      val errors: util.Set[ValidationMessage] = schema.validate(node.toPrettyString, InputFormat.JSON)
+
+      errors.iterator().next().getMessage shouldBe "$.foi_exemption_code: must have at least 1 items but found 0"
+    }
+
+    "pass open document closed when using closureSchema " in {
```
[the remark this answers is not in the document]
Comment (TomJKing) [reply src/test/scala/uk/gov/tna/tdr/metadata/schema/validator/SchemaDataTypeSpec.scala:42]: rename test

File: src/test/scala/uk/gov/tna/tdr/metadata/schema/validator/SchemaDataTypeSpec.scala [hunk 21]
```diff
@@ -0,0 +1,144 @@
...
+import com.networknt.schema._
+import org.scalatest.matchers.should.Matchers._
+import org.scalatest.wordspec.AnyWordSpec
+
+import java.io.{File, InputStream}
+import java.nio.file.{Files, Path}
+import java.util
+
+
+class SchemaDataTypeSpec extends AnyWordSpec {
```
[the remark this answers is not in the document]
Comment (TomJKing) [reply src/test/scala/uk/gov/tna/tdr/metadata/schema/validator/SchemaDataTypeSpec.scala:13]: Add test to cover the above

File: src/test/resources/data/testData.json [hunk 22]
```diff
@@ -0,0 +1,11 @@
+{
```
[the remark this answers is not in the document]
Comment (TomJKing) [reply src/test/resources/data/testData.json:1]: Rename this file

File: src/test/scala/uk/gov/tna/tdr/metadata/schema/validator/SchemaDataTypeSpec.scala [hunk 23]
```diff
@@ -0,0 +1,144 @@
...
+  def transformKey(key: String): String = {
+    val keyMap = Map("FOI decision asserted" -> "foi_exemption_asserted", "Alternative description" -> "description_alternate", "FOI exemption code" -> "foi_exemption_code", "Closure Period" -> "closure_period", "Closure status" -> "closure_type", "Translated title of record" -> "translated_title_of_record", "Add alternative title without the file extension" -> "title_alternate", "Alternative description" -> "description_alternate", "Description" -> "description", "Language" -> "language", "Filename" -> "filename", "Date of the record" -> "date_of_the_record", "Is the description sensitive for the public?" -> "description_public", "Closure Start Date" -> "closure_start_date", "Is the title sensitive for the public?" -> "title_public", "Former reference" -> "former_reference", "Date last modified" -> "date_last_modified", "Closure Start Date" -> "closure_start_date", "Filepath" -> "identifier")
+
+
+    keyMap.getOrElse(key, key)
+
+
+  }
+
+  def transformValue(value: String): Any = {
```
[the remark this answers is not in the document]
Comment (TomJKing) [reply src/test/scala/uk/gov/tna/tdr/metadata/schema/validator/SchemaDataTypeSpec.scala:68]: remove this function if its not needed

[the remark this answers is not in the document]
Comment (ian-hoyle) [reply src/test/scala/uk/gov/tna/tdr/metadata/schema/validator/SchemaDataTypeSpec.scala]: removed

File: metadata-schema/closureSchema.schema.json (hunk 13 above)
[the remark this answers is not in the document]
Comment (ian-hoyle) [reply metadata-schema/closureSchema.schema.json:63]: Three possible Closed spellings as suggested by Emma

File: metadata-schema/relationshipSchema.schema.json (hunk 15 above)
[the remark this answers is not in the document]
Comment (ian-hoyle) [reply metadata-schema/relationshipSchema.schema.json:1]: more relationships added

File: src/test/scala/uk/gov/tna/tdr/metadata/schema/validator/SchemaDataTypeSpec.scala (hunk 16 above)
[the remark this answers is not in the document]
Comment (ian-hoyle) [reply src/test/scala/uk/gov/tna/tdr/metadata/schema/validator/SchemaDataTypeSpec.scala]: gone

**Review by TomJKing — APPROVED** [review]
There are still two comments not addresses, but can live with those, in the interest of just getting these changes in.

File: src/test/scala/uk/gov/tna/tdr/metadata/schema/validator/SchemaDataTypeSpec.scala (hunk 21 above)
[the remark this answers is not in the document]
Comment (ian-hoyle) [reply src/test/scala/uk/gov/tna/tdr/metadata/schema/validator/SchemaDataTypeSpec.scala]: @TomJKing added test for this property with invalid enum value, eg FOI exemption code
expecting change when requirements finalised pattern not to be used but format date so no need for test

**Pull request merged**